File: Count Complete Tree Node.java
// https://leetcode.com/problems/count-complete-tree-nodes/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int countNodes(TreeNode root) {
        
        // If root is null, exit function
        if (root == null) {return 0;}
        
        // If root is by itself, exit function
        if (root.left == null && root.right == null) {return 1;}
        
        // Create 2 new arrays
        ArrayList<TreeNode> children = new ArrayList<TreeNode>();
        ArrayList<TreeNode> queue = new ArrayList<TreeNode>();
        
        // Add root to queue
        queue.add(root);
        
        while (queue.size() > 0) {
            
            // Pop first node from array
            TreeNode node = queue.remove(0);
                
            // If node isn't null, add it to children array
            if (node != null) {
                children.add(node);
            }

            // If node has any children, add it to queue
            if (node.left != null) {
                queue.add(node.left);
            }

            if (node.right != null) {
                queue.add(node.right);
            }

            // Remove node from queue
            queue.remove(node);
        }
        return children.size();
    }
}

File: Count Complete Tree Node.py
# https://leetcode.com/problems/count-complete-tree-nodes/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def countNodes(self, root: Optional[TreeNode]) -> int:
      
        """
          OBJECTIVE: Count the number of nodes there are in the binary tree with a time complexity less than O(n).
          NOTE: I answered the question by using BFS. I used BFS to collect all nodes on the same level.
                The time complexity is O(n) where n = number of nodes inside the tree.
        """
        
        # If root is node is empty, exit function
        if root is None:
            return 0
        
        # If root is by itself, return 1
        if root.left is None and root.right is None:
            return 1
        
        # Create a children and queue list
        children = []
        queue = [root]
        
        while queue:
            
            # Iterate queue list
            for node in queue:
                
                # If node isn't empty, add it to list
                if node != None:
                    children.append(node)
                    
                # If node has any children, add it to queue
                if node.right:
                    queue.append(node.right)
                    
                if node.left:
                    queue.append(node.left)
                
                # Remove node from queue
                queue.remove(node)
                
        return len(children)
